Dereference projections and predicate pushdown in Hive

phd3 opened the pull request:
(required by #1953)

This PR includes the following major changes:

**1. Add support for projected columns in Hive**

* Page sources need to be supplied projected columns representing internal fields from a struct column. We modify `HiveColumnHandle` to include this information about column subfield access. Right now, the `HiveColumnProjectionInfo` simply stores a chain of dereferences, but that can later be extended to support other types of partial columns (eg. subscripts)

* Internal page sources in Hive Connector may or may not be able to push down projected column reads, depending on the formats. Reader page sources or record cursor can provide the partial projections they'll apply, and the rest is "adapted" by the HivePageSource or HiveReaderProjectionsAdaptingRecordCursor.

* By default, the readers only read base columns for projected columns. For example, column "a" will be read for a projection "a.x".  This is implemented through `ReaderProjections#projectBaseColumns`

**2. Pushdown for Dereferences in Hive:**

* The goal is to pushdown dereference expressions to Hive Connector for projections and predicates both.

* As described in #1953, creating partial column handles in `HiveMetadata#applyProjection` will give us both predicate and projection pushdown into Hive through iterative optimization. 

* `HiveMetadata#applyProjection` extracts longest simple dereference chains and variables from input expressions and provies new columnhandles for them. All the columnhandles returned by this method are unique.

**3. Projection and Predicate pushdown of subfields in Columnar Sources:**

* In case of Columnar formats (eg. ORC, Parquet), the partial columns can be read without reading entire columns, which is the primary motive here. 

* We implement p`rojectSufficientColumns` for such page sources to leverage, which lets the pagesource only read the required superset columns. Say we have a column C of type `RowType`. The query asks for a set of subcolumns S from column C. Then the superset columns here is the smallest set of columns from S such that all columns in S can be derived using identity or dereference expressions. i.e when S = {"a.b.c", "a.d" and "a.b.c.x"}, the reader will read create columns {"a.b.c", "a.d"}. The HivePageSource will adapt for ".x" for the third column.

* ORC and Parquet changes are extracted out in different PRs.

martint commented on presto-spi/src/main/java/io/prestosql/spi/expression/ConnectorExpressionVisitor.java:
```diff
@@ -0,0 +1,46 @@
+
+import javax.annotation.Nullable;
+
+public abstract class ConnectorExpressionVisitor<R, C>
```
I intentionally avoided adding a visitor abstraction for ConnectorExpression, as it makes the hierarchy rigid. We'll probably want to make it extensible in the future as we start expanding on the kinds of things that can be communicated across the SPI (e.g.., connector-provided expression types and rules)

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1786,6 +1798,137 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+                .orElseThrow(() -> new TableNotFoundException(tableName));
+
+        // Support ORC tables only
+        if (!extractHiveStorageFormat(table).equals(HiveStorageFormat.ORC)) {
```
The format can vary partition by partition, so it needs to be supported for all readers.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java:
```diff
@@ -405,6 +435,66 @@ public int getIndex()
+            checkState(!loaded, "Already loaded");
+            Block currentBlock = parentBlock;
+            for (int i = 0; i < dereferences.size(); i++) {
+                currentBlock = toColumnarRow(currentBlock).getField(dereferences.get(i));
```
@wagnermarkd `toColumnarRow` method uses lazyBlock.getBlock() in case the input block is a lazy block. After changes in #1629 , `LazyBlock::getBlock`, invocation does not load the underlying blocks recursively. (unlike `getLoadedBlock`)   hence the `toColumnarRow` invocation does not load everything. the last `getLoadedBlock` call loads the required dereferenced block. 

The "adapatation" here is about transforming the block to correspond to the required nested type, rather than making it lazy.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java:
```diff
@@ -154,6 +161,28 @@ public ConnectorPageSource createPageSource(ConnectorTransactionHandle transacti
+            }
+        }
+
+        DereferenceAdaptation dereferenceAdaptation = new DereferenceAdaptation(hiveNestedColumns, baseColumnIndexToChannelIndex);
```
@wagnermarkd I agree with the first point, that pushing dereferences into readers sets the stage for more optimizations. I can add that in this PR.

I believe that recent change in to LazyBlock already achieves (2).  https://github.com/dain/presto/blob/5a15b5189421d493471c3a5de20a2fca90b0a079/presto-spi/src/main/java/io/prestosql/spi/block/LazyBlock.java#L351

wagnermarkd replied:
Talked offline. You're right: the recent changes to the Orc reader/block loading only load the top level block, and toColumnarRow doesn't load the fields.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1800,6 +1809,92 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+        // Modify projections to refer to new variables
+        List<ConnectorExpression> newProjections = projections.stream()
+                .map(expression -> replaceWithNewVariables(expression, expressionToVariableMappings.build()))
+                .collect(toImmutableList());
```
`replaceWithNewVariables` and `getSupportedSubExpressions` methods do per-expression processing. The loop that goes over `supersetSubExpressions` to create new assignments also does that one assignment at a time. For `getSupersetSubExpressions`, we do need all the context about  so it needs to be done with all expressions as an argument. Can you elaborate more on what you mean by one assignment at a time?

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveColumnHandle.java:
```diff
@@ -65,47 +68,98 @@
+            @JsonProperty("baseColumnName") String baseColumnName,
+            @JsonProperty("baseHiveColumnIndex") int baseHiveColumnIndex,
+            @JsonProperty("baseColumnHiveType") HiveType baseColumnHiveType,
+            @JsonProperty("baseColumnType") Type baseColumnType,
```
Yes, I'll add some comments. Was also thinking that we can move the `hiveType` and `type` out of the constructor arguments, since they can be derived from other info during the construction. what do you think?

findepi commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1800,6 +1809,92 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+                .flatMap(expression -> getSupportedSubExpressions(expression).stream())
+                .collect(toImmutableSet());
+
+        // Get a superset of expressions. eg. superset of expressions {"a.b", "a.b.c"} is {"a.b"}.
```
@wagnermarkd does HMS store stats for struct's individual fields?

wagnermarkd replied:
@findepi, last I looked it did not.

phd3 commented:
@wagnermarkd split the PR into two parts. #1868 deals with the engine side changes.

phd3 commented:
@wagnermarkd @martint this PR is ready for review again.

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePartialColumnAdapter.java:
```diff
@@ -0,0 +1,252 @@
+
+public class HivePartialColumnAdapter
+{
+    public enum PartialColumnLoadingStrategy
```
(Based on our discussion the other day...)

From an abstraction perspective, I think it might be cleaner to approach it this way:
* The connector describes what dereference paths it wants from the reader upon instantiating it
* The instantiation returns a reader + a list of (possibly incomplete) paths it's able to satisfy
* The connector inserts an adaptation layer around the reader if necessary to satisfy anything the reader is not able to produce

For example, let's say the connector promised the engine it will produce `a.x`, `a.y` and `b`. When it instantiates the reader, it provides `[a.x, a.y, b]` as its preferences. Let's suppose the reader can only read full columns. In such a case, it would return `[a, b]` as the list of what it can actually provide. The connector then inserts and adapter from `a` -> `[a.x, a.y]`.

I think this might fit well with other kinds of adaptations we might need to do, such as for type coercions. The connector might ask the reader to return something as a BIGINT, but the reader would say "I only support INTEGER". Then the connector would insert an adapter to convert the types.

martint replied:
@wagnermarkd, how does this fit within the work you've been doing around schema evolution?

phd3 commented:
High level changes after reviews from @wagnermarkd and @martint 

* Reference columns with dereferences as "projected" columns instead of "partial" columns. This dereference pushdown is the first step towards pushing different kinds of operations to the connector.  "projected" columns seems like a more generic and accurate name than partial columns.

* Allow `HivePageSourceFactory` and `HiveRecordCusorProvider` to return supported projected columns. As @martint mentioned, HivePageSourceProvider will provide what dereference paths it wants from the reader upon creating it, which returns a reader + a list of (possibly incomplete) paths it's able to satisfy. HivePageSourceProvider is responsible for creating either (1) a HivePageSource with required adaptations or (2) a HiveReaderProjectionsAdaptingRecordCursor. 

* `ReaderProjections` store information for mapping columns required by Hive to corresponding "partially satisfied" columns given by the reader.

* `ReaderProjectionsAdapter` adapts a page, `HiveReaderProjectionsAdaptingRecordCursor` adapts a record cursor. 

* `applyProjection` is roughly the same module renamings.

* Testing:
- Unit tests with `TestReaderProjections`, `TestReaderProjectionsAdapter`. 
- Tests for pagesources and recordcursors in TestFileFormats
- Tests for `HiveMetadata::applyProjection` in AbstractTestHive
- end-to-end tests in `TestHiveIntegrationSmokeTest`
- looking into how we can test projection and predicate pushdown together with something like #2125.

phd3 commented on presto-spi/src/main/java/io/prestosql/spi/block/AbstractRowBlock.java:
```diff
@@ -174,6 +174,10 @@ public Block copyRegion(int position, int length)
         }
         checkReadablePosition(position);
 
+        if (isNull(position)) {
```
hmm, looks like this method shouldn't be invoked if `isNull` was true for the position. Removed.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/GenericHiveRecordCursorProvider.java:
```diff
@@ -73,6 +73,10 @@ public GenericHiveRecordCursorProvider(HdfsEnvironment hdfsEnvironment, DataSize
             TypeManager typeManager,
             boolean s3SelectPushdownEnabled)
     {
+        if (s3SelectPushdownEnabled) {
```
My understanding is that when `s3SelectPushdownEnabled` is enabled and serde is present in `CSV_SERDES`,  `S3SelectRecordCursor` should be picked over `GenericHiveRecordCursor`. Right now in the code, this choice depends on the implicit iteration order of `Set<HiveRecordCursorProvider> cursorProviders`. i.e. if the iteration ordered picked `GenericHiveRecordCursorProvider` first, then it would be the one to create the cursor. The purpose of adding this commit was to make the preference explicit. 

But I realized that in cases when `s3SelectPushdownEnabled` is true and the serde is not in `CSV_SERDES`, the current code falls back to using GenericHiveRecordCursor.  The commit I added changes that. That's why I've removed this commit, since the code works right now. But I think we shouldn't rely on the set's iteration order. What's your take here?

phd3 commented:
Added a benchmark in #2307

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjectionsAdapter.java:
```diff
@@ -0,0 +1,245 @@
+
+        Block[] blocks = new Block[outputToInputMapping.size()];
+
+        // Prepare adaptaions to extract dereferences
```
Typo: "adaptaions"

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjectionsAdapter.java:
```diff
@@ -0,0 +1,245 @@
+    private static class DereferenceBlockLoader
+            implements LazyBlockLoader
+    {
+        boolean loaded;
```
`private`

Also, move the final fields before the mutable ones (no strict convention about this, but we generally tend to organize them that way).

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjectionsAdapter.java:
```diff
@@ -0,0 +1,245 @@
+    {
+        boolean loaded;
+        private Block inputBlock;
+        private final List<Integer> dereferenceSequence;
```
How does this work for "by name" semantics?

phd3 replied:
https://github.com/prestosql/presto/pull/1720#issuecomment-568143233

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjectionsAdapter.java:
```diff
@@ -0,0 +1,245 @@
+            }
+
+            return newlyCreatedBlock.build();
+        }
```
Hmm... this is not ideal. There will be a noticeable performance hit due to the extra copying and assembly. What we need is a way to assemble a block for this column with the "null vector" from the parent row block and block for the field. @dain, any thoughts on this?

phd3 replied:
@martint agreed. I believe we'll need to assemble a block by reindexing, considering nulls. similar to https://github.com/prestosql/presto/blob/ed99fb27e35605da80fda28f1b5c10424e3bc53a/presto-spi/src/main/java/io/prestosql/spi/block/RowBlock.java#L48

I don't know if there is any existing Block implementation that provides this. `DictionaryBlock` won't work, because its implementation does not allow representing nulls through negative indexes. If we modify that behavior, it'd probably require changing accessors of `DictionaryBlock`. Or we could create a new block implementation altogether for this purpose.

phd3 replied:
@martint applying dereference projection in Hive connector means that we're getting rid of this projection in the engine (likely in SFP operator). We do have the same issue of copying data there as well, right? How do we resolve it right now in the engine?

martint replied:
Yes, you're right. It loops over the values, gets the field for each row element and appends to a new block. Ok, this is probably fine for now, but there's an opportunity to improve performance later. Let's make a note of it in a TODO.

phd3 replied:
I've added a TODO for `adaptNulls` method.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjections.java:
```diff
@@ -0,0 +1,104 @@
+    {
+        this.readerColumns = ImmutableList.copyOf(requireNonNull(readerColumns, "readerColumns is null"));
+
+        readerBlockIndices.stream().allMatch(value -> value >= 0 && value < readerColumns.size());
```
oh, right. Changing it to `forEach(x --> checkArgument(...))`

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveReaderProjectionsAdaptingRecordCursor.java:
```diff
@@ -0,0 +1,239 @@
+        // Apply all dereferences except for the last one, since the last one requires getBoolean invocation
+        for (int j = 0; j < dereferences.size() - 1; j++) {
+            int dereferenceIndex = dereferences.get(j);
+            baseObject = baseObject.getObject(dereferenceIndex, Block.class);
```
that check happens in `isNull` method, so I think we don't need to worry about it here.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java:
```diff
@@ -397,6 +468,29 @@ public int getIndex()
+        EMPTY
+    }
+
+    private static Optional<BucketAdaptation> createBucketAdaptation(Optional<BucketConversion> bucketConversion, OptionalInt bucketNumber, List<ColumnMapping> columnMappings)
```
we also filter out projection (non-base) columns while constructing `baseHiveColumnToBlockIndex` . Apart from that it's the two changes you mentioned.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java:
```diff
@@ -334,37 +344,86 @@ public int getIndex()
             return columnMappings.build();
         }
 
+        private static boolean projectionExistsInPartitionSchema(Optional<HiveType> baseTypeCoercionFrom, HiveColumnHandle column)
```
a struct column in the partition schema is coercible with the table column, even if subfields are added/removed at the end. This method checks if the "baseTypeCoercionFrom" has the dereference chain specified in the input "column".

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjections.java:
```diff
@@ -0,0 +1,104 @@
+    /**
+     * Creates unique base columns for input projected columns and prepares a mapping between the two.
+     */
+    public static Optional<ReaderProjections> projectBaseColumns(List<HiveColumnHandle> columns)
```
rephrased

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveReaderProjectionsAdaptingRecordCursor.java:
```diff
@@ -0,0 +1,239 @@
+            int dereferenceIndex = dereferences.get(j);
+            baseObject = baseObject.getObject(dereferenceIndex, Block.class);
+            baseType = baseType.getTypeParameters().get(dereferenceIndex);
+        }
```
Good points. done.

phd3 commented:
Thanks for bringing up the name-based field access issue. I believe that Hive side changes work as expected, but pagesources would need to use names while getting the nested columns. (I've added that change to the ORC pushdown commit). I’ve also added a test with these examples which succeeds before and after the change in this PR: https://github.com/prestosql/presto/pull/1720/commits/b7de259cb1cc7b241c70788967929e8ef4793fb0   

To elaborate on this, please consider the following examples.

* A row-type column is only considered coercible between partition and table if subfields are added/dropped at the end. Renames/Reordering is not allowed in Presto currently. This works as before.

projection | tablecolumn | partitioncolumn | filecolumn | before | after #1720
-- | -- | -- | -- | -- | --
“a.b” | a(b, c) | a(c, b) | don’tcare | Failure | Failure

* If table/partition schemas are coercible, it means that partitions have the same partial order and names of subfields. It is possible that partition may not have the desired subfields. In this case, the dereference would return null, irresepctive of whether the file has those subfields or not. 

projection | tablecolumn | partitioncolumn | filecolumn | before | after #1720
-- | -- | -- | -- | -- | --
“a.c” | a(b, c) | a(b) | a(b, c) | null | null

* When table/partition columns match exactly,  but the file has a different order of subfields. Then dereference access must correctly read subfields for ORC/Parquet (and even AVRO I believe) since it’s name based. This works currently in Presto, since _OrcPageSource uses the **readType** provided by the HivePageSourceProvider, which is the partition's type_. So in case where the partition expects a(b, c) but the file contains a(c, b), the ORCPageSource still reads the column as a(b, c). I’ve modified ORC pushdown to make sure that the same behaviour is supported with the projected columns. The following table is true for ORC and Parquet.

projection | tablecolumn | partitioncolumn | filecolumn | before | after #1720
-- | -- | -- | -- | -- | --
“a.b” | a(b, c) | a(b, c) | a(c, b) | reads value | reads value
“a.c.y”, “a.c” | a(b, c(x, y)) | a(b, c(x, y)) | a(c(y, x), b) | reads value | Reads correctly in reader + adapts correctly in HivePageSource

* `useOrcColumnNames` and `useParquetColumnNames` work as expected like before, since the logic corresponding to base columns has not changed and these properties only impact at the base columns.

Do you know if we're missing any cases here?

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java:
```diff
@@ -152,14 +157,22 @@ public HivePageSource(
-            if (columnMapping.getCoercionFrom().isPresent()) {
-                coercers.add(Optional.of(createCoercer(typeManager, columnMapping.getCoercionFrom().get(), columnMapping.getHiveColumnHandle().getHiveType())));
+            if (columnMapping.getKind() != EMPTY && columnMapping.getBaseTypeCoercionFrom().isPresent()) {
+                List<Integer> dereferenceIndices = column.getHiveColumnProjectionInfo()
```
https://github.com/prestosql/presto/pull/1720#issuecomment-568143233

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java:
```diff
@@ -377,16 +436,28 @@ public int getIndex()
+                        HiveType fromHiveTypeBase = columnMapping.getBaseTypeCoercionFrom().get();
+
+                        Optional<HiveColumnProjectionInfo> newColumnProjectionInfo = columnHandle.getHiveColumnProjectionInfo().map(projectedColumn -> {
+                            HiveType fromHiveType = fromHiveTypeBase.getHiveTypeForDereferences(projectedColumn.getDereferenceIndices()).get();
```
https://github.com/prestosql/presto/pull/1720#issuecomment-568143233

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java:
```diff
@@ -334,37 +344,86 @@ public int getIndex()
+
+            HiveType partitionType = baseTypeCoercionFrom.get();
+            List<Integer> dereferences = column.getHiveColumnProjectionInfo()
+                    .map(projectedColumn -> projectedColumn.getDereferenceIndices())
```
https://github.com/prestosql/presto/pull/1720#issuecomment-568143233

martint commented:
> useOrcColumnNames and useParquetColumnNames work as expected like before, since the logic corresponding to base columns has not changed and these properties only impact at the base columns.
> Do you know if we're missing any cases here?

How does Hive behave in this case for ORC/Parquet. The current behavior in Presto is not necessarily what we want in the long term. When querying Hive data, we try to be consistent with how Hive does it, but we may have a couple of bugs.

In particular, what happens in Hive if a table has a column of type `row(a bigint, b bigint)`, a file declares the column type as `row(b bigint, a bigint)` and a query references `col.a`?

martint commented:
> In particular, what happens in Hive if a table has a column of type row(a bigint, b bigint), a file declares the column type as row(b bigint, a bigint) and a query references col.a?

Another thought about how to reason about dereference semantics: if the dereference is handled in the engine, how does it behave? Of course, this will depend on how data is laid out in the row values coming out of the connector. If a partition has the columns in a different order or different names, how do they get presented to the engine (vis-a-vis the table schema)?

phd3 commented:
> In particular, what happens in Hive if a table has a column of type `row(a bigint, b bigint)`, a file declares the column type as `row(b bigint, a bigint)` and a query references `col.a`?

@martint Hive reads the correct value for the field from the file.

phd3 commented:
> Another thought about how to reason about dereference semantics: if the dereference is handled in the engine, how does it behave? Of course, this will depend on how data is laid out in the row values coming out of the connector. If a partition has the columns in a different order or different names, how do they get presented to the engine (vis-a-vis the table schema)?

- partition is allowed to have different order for columns, but not for subfields inside a row column. this is something we can consider during the schema evolution changes.

- When the data comes out of the reader (OrcPageSource), it matches the partition schema (the reader takes care of the subfield reordering). 
- when the data comes out of the connector (HivePageSource), it matches the schema defined for the table. 

- that's why dereference in the engine only cares about what is sees in the table level schema. Column level reordering is taken care of in the connector.

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveReaderProjectionsAdaptingRecordCursor.java:
```diff
@@ -184,7 +184,7 @@ public Object getObject(int field)
 
         if (dereferences.isEmpty()) {
-            return delegate.getSlice(inputFieldIndex);
+            return delegate.getObject(inputFieldIndex);
```
Why did this change? If this was a bug in a previous commit, apply it there, instead.

martint replied:
I think you forgot to address this comment.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSourceProvider.java:
```diff
@@ -334,37 +344,86 @@ public int getIndex()
 
+        private static boolean projectionExistsInPartitionSchema(Optional<HiveType> baseTypeCoercionFrom, HiveColumnHandle column)
+        {
+            if (!baseTypeCoercionFrom.isPresent()) {
```
I see what you mean. I've taken the `baseTypeCoercionFrom.isPresent()` check out since it's the semantics of the caller method. also renamed the method to `projectionExistsForType`.

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveApplyProjectionUtil.java:
```diff
@@ -0,0 +1,225 @@
+        return false;
+    }
+
+    static ProjectedColumnRepresentation createProjectedColumnRepresentation(ConnectorExpression expression)
```
`createOrderPrefixes` method made more sense earlier when it was being used at multiple places, replaced with the suggested implementation

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveApplyProjectionUtil.java:
```diff
@@ -0,0 +1,225 @@
+            }
+        }
+
+        return false;
```
compacted this into `isProjectionSupported` method

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveApplyProjectionUtil.java:
```diff
@@ -0,0 +1,225 @@
+
+        if (expression instanceof Constant) {
+            return expression;
+        }
```
I had intentionally kept it for readability, like switch/case statements so that it's easier to add logic for future additions to `ConnectorExpression` family. But it's not critical, removed it in the update.

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java:
```diff
@@ -283,6 +312,28 @@ else if (column.getBaseHiveColumnIndex() < fileColumns.size()) {
+
+        OrcColumn current = baseOrcColumn;
+        List<String> dereferenceNames = partialColumnInfo.get().getDereferenceNames();
+        for (int i = 0; i < dereferenceNames.size(); i++) {
```
This can be a for-each loop:

```java
for (String field : partialColumnInfo.get().getDereferenceNames()) {
```

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java:
```diff
@@ -283,6 +312,28 @@ else if (column.getBaseHiveColumnIndex() < fileColumns.size()) {
+        List<String> dereferenceNames = partialColumnInfo.get().getDereferenceNames();
+        for (int i = 0; i < dereferenceNames.size(); i++) {
+            // Get nested columns by name
+            Map<String, OrcColumn> nestedColumns = current.getNestedColumns().stream()
```
No point in building a map to do a single lookup. Iterate over the list manually and find the first element that matches:

```java
OrcColumn current = baseOrcColumn;
for (String field : partialColumnInfo.get().getDereferenceNames()) {
    Optional<OrcColumn> orcColumn = current.getNestedColumns().stream()
            .filter(column -> column.getColumnName().toLowerCase(ENGLISH).equals(field))
            .findFirst();

    if (!orcColumn.isPresent()) {
        return null;
    }

    current = orcColumn.get();
}
```

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java:
```diff
@@ -208,44 +210,71 @@ private static OrcPageSource createOrcPageSource(
+                }
+
+                if (baseOrcColumn != null) {
+                    OrcColumn orcColumn = getNestedOrcColumn(baseOrcColumn, column.getHiveColumnProjectionInfo());
```
@martint I separated the predicate-building from columnAdaptations-building thinking of the possibility that `effectivePredicate` contains a column which is not present in `columns`. 

I don't think this happens right now, because effectivePredicate is still re-tested inside the plan. Can you please confirm that?

The separate computation of predicates would be required if the following state is possible.

```
TableScan
     assignments -->  x := column(a.b)
     effectivePredicate --> a = row(2, 3)        
```

But I don't think this can happen right now, assignments will have column `a` as well, since it's in the filter. Please let me know if we're missing any cases here. Otherwise it should be okay to go with your suggestion.

martint replied:
Correct. The predicate has to reference columns listed in the assignments.

phd3 commented:
@martint addressed your comments and updated the PR.

martint commented on presto-hive/src/test/java/io/prestosql/plugin/hive/optimizer/TestPushProjectionRuleWithHive.java:
```diff
@@ -0,0 +1,159 @@
+        PushProjectionIntoTableScan pushProjectionIntoTableScan = new PushProjectionIntoTableScan(tester().getMetadata(), tester().getTypeAnalyzer());
+
+        tester().getQueryRunner().execute("CREATE TABLE " +
+                HIVE_CATALOG_NAME + "." + SCHEMA_NAME + "." + TABLE_NAME + "\n" +
```
Why qualified name?

Also, format as per the suggestion in the other test.

phd3 commented:
@martint thanks for your latest comments. I think it'd be easier to address them once #2125 is checked-in to avoid duplicate reviews).

phd3 commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjections.java:
```diff
@@ -101,4 +106,116 @@ public int readerColumnPositionForHiveColumnAt(int index)
+     * Creates a set of sufficient columns for the input projected columns and prepares a mapping between the two. For example,
+     * if input {@param columns} include columns "a.b" and "a.b.c", then they will be projected from a single column "a.b".
+     */
+    public static Optional<ReaderProjections> projectSufficientColumns(List<HiveColumnHandle> columns)
```
we use it for implementing pushdown in ORC format. looks like it got dropped by mistake while rebasing. updated in the latest push.

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjections.java:
```diff
@@ -0,0 +1,221 @@
+        return Optional.of(new ReaderProjections(sufficientColumns, outputColumnMapping.build()));
+    }
+
+    static class DereferenceChain
```
I think this class can be private

martint commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -3181,6 +3181,81 @@ public void testBucketFilteringByInPredicate()
+            assertUpdate("INSERT INTO evolve_test values (1, row(1), 1)", 1);
+            assertUpdate("ALTER TABLE evolve_test DROP COLUMN a");
+            assertUpdate("ALTER TABLE evolve_test ADD COLUMN a row(b bigint, c varchar)");
+            //assertUpdate("INSERT INTO evolve_test values (2, row(2, 'def'), 2)", 1);
```
Why is this commented out?

martint commented on presto-hive/src/test/java/io/prestosql/plugin/hive/orc/TestOrcPageSourceFactory.java:
```diff
@@ -0,0 +1,218 @@
+        ImmutableList.Builder<String> splitPropertiesColumnNames = ImmutableList.builder();
+        ImmutableList.Builder<String> splitPropertiesColumnTypes = ImmutableList.builder();
+        Set<String> baseColumnNames = new HashSet<>();
+        for (int i = 0; i < columnsToRead.size(); i++) {
```
This can be a foreach loop

martint commented:
There are failures in produce tests. I think this may be what's causing them:

```
 io.prestosql.spi.PrestoException: Error opening Hive split hdfs://hadoop-master:9000/user/hive/warehouse/test_full_acid_table_read/part_col=2/base_0000001/bucket_00001 (offset=0, length=668): readColumns contains duplicate entries
ptl-presto-master   | 	at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.createOrcPageSource(OrcPageSourceFactory.java:337)
ptl-presto-master   | 	at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.createPageSource(OrcPageSourceFactory.java:154)
ptl-presto-master   | 	at io.prestosql.plugin.hive.HivePageSourceProvider.createHivePageSource(HivePageSourceProvider.java:180)
ptl-presto-master   | 	at io.prestosql.plugin.hive.HivePageSourceProvider.createPageSource(HivePageSourceProvider.java:111)
ptl-presto-master   | 	at io.prestosql.plugin.hive.HivePageSourceProvider.createPageSource(HivePageSourceProvider.java:94)
ptl-presto-master   | 	at io.prestosql.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:49)
ptl-presto-master   | 	at io.prestosql.split.PageSourceManager.createPageSource(PageSourceManager.java:63)
ptl-presto-master   | 	at io.prestosql.operator.TableScanOperator.getOutput(TableScanOperator.java:288)
ptl-presto-master   | 	at io.prestosql.operator.Driver.processInternal(Driver.java:379)
ptl-presto-master   | 	at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
ptl-presto-master   | 	at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
ptl-presto-master   | 	at io.prestosql.operator.Driver.processFor(Driver.java:276)
ptl-presto-master   | 	at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1075)
ptl-presto-master   | 	at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
ptl-presto-master   | 	at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
ptl-presto-master   | 	at io.prestosql.$gen.Presto_330_96_gdbfb462____20200225_221503_2.run(Unknown Source)
ptl-presto-master   | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
ptl-presto-master   | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
ptl-presto-master   | 	at java.lang.Thread.run(Thread.java:748)
ptl-presto-master   | Caused by: java.lang.IllegalArgumentException: readColumns contains duplicate entries
ptl-presto-master   | 	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:141)
ptl-presto-master   | 	at io.prestosql.orc.OrcRecordReader.<init>(OrcRecordReader.java:145)
ptl-presto-master   | 	at io.prestosql.orc.OrcReader.createRecordReader(OrcReader.java:275)
ptl-presto-master   | 	at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.createOrcPageSource(OrcPageSourceFactory.java:296)
ptl-presto-master   | 	... 18 more
```

tooptoop4 commented:
@phd3 just 1 conflict left

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/rcfile/RcFilePageSourceFactory.java:
```diff
@@ -152,7 +160,8 @@ else if (deserializerClassName.equals(ColumnarSerDe.class.getName())) {
                     DataSize.of(8, Unit.MEGABYTE));
 
-            return Optional.of(new RcFilePageSource(rcFileReader, columns));
+            ConnectorPageSource rcFilePageSource = new RcFilePageSource(rcFileReader, projectedReaderColumns);
```
The variable could be named just `pageSource` for conciseness.

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/s3select/S3SelectRecordCursorProvider.java:
```diff
@@ -80,12 +83,22 @@ public S3SelectRecordCursorProvider(HdfsEnvironment hdfsEnvironment, PrestoS3Cli
 
+        Optional<ReaderProjections> projectedReaderColumns = projectBaseColumns(columns);
+        // Ignore predicates on partial columns for now.
+        effectivePredicate = effectivePredicate.transform(x -> x.isBaseColumn() ? x : null);
```
```suggestion
        effectivePredicate = effectivePredicate.transform(column -> column.isBaseColumn() ? column : null);
```

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveApplyProjectionUtil.java:
```diff
@@ -0,0 +1,185 @@
+
+final class HiveApplyProjectionUtil
+{
+    private HiveApplyProjectionUtil(){}
```
```suggestion
    private HiveApplyProjectionUtil() {}
```

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjections.java:
```diff
@@ -103,4 +108,115 @@ public int readerColumnPositionForHiveColumnAt(int index)
+    {
+        requireNonNull(columns, "columns is null");
+
+        if (columns.stream().allMatch(column -> column.isBaseColumn())) {
```
```suggestion
        if (columns.stream().allMatch(HiveColumnHandle::isBaseColumn)) {
```

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveApplyProjectionUtil.java:
```diff
@@ -102,6 +102,10 @@ public static ConnectorExpression replaceWithNewVariables(ConnectorExpression ex
 
     /**
      * Returns the assignment key corresponding to the column represented by {@param projectedColumn} in the {@param assignments}, if one exists.
+     * The variable in the {@param projectedColumn} can itself be a representation of another projected column. For example,
```
Does this comment reflect new behavior? If not, move it to the commit that introduced the method.

martint commented on presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:
```diff
@@ -1868,10 +1868,14 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
+            ConnectorSession session,
+            ConnectorTableHandle handle,
+            List<ConnectorExpression> projections,
+            Map<String, ColumnHandle> assignments)
```
Since this is just a formatting change of the original code, move to the commit that introduced `applyProjection`

phd3 commented:
@martint Thanks for the review, addressed your comments. 

There're two main other changes during the rebase:

1. Excluding test for applyProjection from alluxio metastore since it involves creating a table
2. Resolving conflicts to use TableToPartitionMapping introduced instead of coercions in HivePageSourceProvider::buildColumnMappings.
